feat: flatten chained map, forEach and reduce for vtl

thantos opened the pull request:
When a map precedes a `reduce`, `forEach`, or a `map`, apply the map transformations inline in the same VTL foreach.

Example
```ts
reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list
        .map((item) => `hello ${item}`)
        .map((item) => `hello ${item}`);
    })
```

Before:

```vtl
#set($v1 = [])
#set($v2 = [])
#foreach($item in $context.arguments.list)
#set($v3 = "hello \${item}")
$util.qr($v2.add($v3))
#end
#foreach($item in $v2)
#set($v4 = "hello \${item}")
$util.qr($v1.add($v4))
#end
...
```

After

```vtl
#set($v1 = [])
#foreach($item in $context.arguments.list)
#set($item = "hello \${item}")
#set($v2 = "hello \${item}")
$util.qr($v1.add($v2))
#end
```

Caveat: When the map, reduce, or foreach uses the array parameter, do not flatten.

Bug: Fixed a bug in reduce where the internal `previous` variable name was initialized outside of the VTL foreach and could override a later reduce which also uses the same variable name.

Example:

```ts
    reflect((context: AppsyncContext<{ list: string[] }>) => {
      return context.arguments.list
        .map((item) => `hello ${item}`)
        .reduce((newList: string[], item) => {
          return [...newList, item];
        }, [])
        .map((item) => `hello ${item}`)
        .reduce((newList: string[], item) => {
          return [...newList, item];
        }, []);
    })
```

Before (generated variable names may be wrong, sorry)
```vtl
#set($newList = [])
#foreach($item in $context.arguments.list)
#set($item = "hello \${item}")
#set($v3 = [])
$util.qr($v3.addAll($newList))
$util.qr($v3.add($item))
#set($v2 = $v3)
#set($v1 = $v2)
#end
#set($newList = [])
#foreach($item in $newList)
#set($item = "hello \${item}")
#set($v6 = [])
$util.qr($v6.addAll($newList))
$util.qr($v6.add($item))
#set($v5 = $v6)
#set($newList = $v5)
#end
```

After
```vtl
#set($v1 = [])
#foreach($item in $context.arguments.list)
#set($item = "hello \${item}")
#set($newList = $v1)
#set($v3 = [])
$util.qr($v3.addAll($newList))
$util.qr($v3.add($item))
#set($v2 = $v3)
#set($v1 = $v2)
#end
#set($v4 = [])
#foreach($item in $v1)
#set($item = "hello \${item}")
#set($newList = $v4)
#set($v6 = [])
$util.qr($v6.addAll($newList))
$util.qr($v6.add($item))
#set($v5 = $v6)
#set($v4 = $v5)
#end
```

Closes #2

sam-goodwin commented:
Thanks for the change, this is great! Do you have some snippets of code to show the improvement?

sam-goodwin commented on src/vtl.ts:
```diff
@@ -418,4 +437,78 @@ export class VTL {
+    return fn.parameters.map((p) => p.name);
+  }
+
+  private renderMapOrForEach(
```
Some docs would be helpful for understanding its purpose.

thantos commented:
> Thanks for the change, this is great! Do you have some snippets of code to show the improvement?

I added more test cases and updated the existing ones.

thantos commented:
Though I could add some in the PR. Will do that.

sam-goodwin commented on src/vtl.ts:
```diff
@@ -418,4 +437,78 @@ export class VTL {
     return __exhaustive;
   }
+
+  private mapForEachArgs(call: CallExpr) {
```
Should this be called `getForEachArgNames`? Does it need to be a function on the class? Could it be static function?
```ts
/**
 * Returns the [value, index, array] arguments if this CallExpr is a `forEach`, `map` or `reduce` call.
 */
export function getForEachArgNames(call: CallExpr): string[] {

}
```

thantos replied:
Yeah, it could be static, will rename too.

sam-goodwin commented on src/vtl.ts:
```diff
@@ -418,4 +437,78 @@ export class VTL {
+    array?: string
+  ) {
+    if (index) {
+      this.add(`#set($${index} = $foreach.index)`);
```
Why `$${index}` instead of `${index}`? I have been trying to adopt the convention of having variables always contain the `$`.

thantos replied:
I didn't change this, index and array were unformatted in your code too. I can update them though.

sam-goodwin replied:
Oh my bad. I changed my mind a few times, left it a bit messy it seems.

If it's easy to do so, please update. No big deal tho.

sam-goodwin commented on src/vtl.ts:
```diff
@@ -418,4 +437,78 @@ export class VTL {
+   *
+   * @return [firstVariable, list variable, render function]
+   */
+  private divePropAccessMapOperations(
```
The name of this function doesn't immediately tell me what it does.

thantos replied:
uhh, yeah, will think of a new name...

sam-goodwin commented on src/vtl.ts:
```diff
@@ -418,4 +437,78 @@ export class VTL {
+        lastValue,
+        list,
+        () => {
+          lastRender();
```
I'm confused how this function works. What does this callback do?

thantos replied:
uhh, yeah, I don't want to apply the `maps` to the VTL until all of them have been resolved. This function returns the variable name to put into the vtl `foreach`, so I want to find the deepest map, build the function to add the map logic to the vtl, return the variable to put in the foreach, and then execute (render) the maps into the `vtl` later. 

I would have done this differently if this class wasn't so stateful. haha

Again, will document.

sam-goodwin replied:
Why do you need a callback to be returned? Can you pass a flag into the recursive call to change how it renders instead of calling back?

thantos replied:
No, or at least I don't want to. This supports flattening maps for reduce and map/foreach (and maybe filter?), but this method is completely decoupled from the top level reduce/map logic including the `foreach` (map, reduce) and `initial value` (reduce) logic.

I didn't think of the function as a callback, but a closure to control when we render (similar to the body() method in reduce).

sam-goodwin commented on src/vtl.ts:
```diff
@@ -418,4 +437,78 @@ export class VTL {
+      const [lastValue, list, lastRender] = this.divePropAccessMapOperations(
+        next,
+        `$${value}`,
+        !!array
```
What does this achieve?

thantos replied:
Will document. A map, reduce, or forEach that uses `array` cannot be simplified because the array wouldn't be accurate to the point in time as `array` is the previous array, which when flattened, has yet to be computed.

sam-goodwin commented on src/vtl.ts:
```diff
@@ -418,4 +437,78 @@ export class VTL {
+      ];
+    }
+    // If the expression isn't a map, return the expression and return variable, render nothing
+    return [returnVariable, this.eval(expr), () => {}];
```
`() => {}`?

thantos replied:
The base case shouldn't do anything, can change it to be undefined.

sam-goodwin replied:
This smells a bit. Makes me think the callback is unnecessary. Base case should be a return from a recursive call instead of an empty callback?

thantos replied:
The "callbacks" give the caller control of when the computed content is applied to the mutable state. Its not a callback, but a renderer trigger. 

Another option is to pass a callback that contains the "before" values (the foreach and other things) into this method, but I like that less because its more of a callback pattern than the late rendering closure. 

And the other option is to build an array of VTL statements instead of mutating the cental statements array, then this wouldn't be an issue because the computed statements could be returned and injected in where needed.

sam-goodwin replied:
I thought the recursive function would recurse until it finds the terminal case. Terminal case would start the `#foreach` and return a string pointing to the result of its computation. The recursive cases would then emit their statements pointing at that.

thantos replied:
I'll give it a try.

I was trying to keep the method independent of the calling implementation, where the foreach is part of the parent, not the flattened maps. 

Looking at the outcome, it appears the only part that comes before the render is the foreach, so maybe it's fine. It'll be a bit strange because the foreach will be in the method, but the end will be outside.

sam-goodwin commented on test/vtl.test.ts:
```diff
@@ -333,18 +333,61 @@ test("chain map over list", () =>
+          const x = i + 1;
+          return `hello ${item} ${x} ${arr.length}`;
+        })
+        .map((item2, ii) => `hello ${item2} ${ii}`);
```
what happens if you name this `i` instead of `ii`?

sam-goodwin replied:
Will it conflict with the other chain calls?

thantos replied:
It's fine because 1. It would be the same value, map can't change i 2. It would overwrite the previous one, 3. The ts compiler will ensure there isn't a conflict.

We may be able to optimize later by proving that a set doesn't make a meaningful change, but just a non functional optimization.

thantos replied:
The previous test shows what happens.
